### PR TITLE
Change Password functionality added

### PR DESCRIPTION
**Pull request by imrahulmane**

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -8,26 +8,32 @@ import { JwtService } from '@nestjs/jwt';
...
 
     async signUp(data: UserDto): Promise<any> {
         const {username, email, password} = data;
 
-        const user = await this.userService.findByEmail(email);
-        
+        const user = await this.userService.findByEmail(email, ['id', 'email', 'password']);
+       
         if(user){
             throw new BadRequestException("Email already in use");
```
Comment (DINESHKARPE) [inline src/auth/auth.service.ts]: Load All Error message from property file

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -50,8 +56,64 @@ export class AuthService {
             throw new UnauthorizedException("Wrong Credentials");
         }
 
-        return this.generateToken( user.id, user.username,);
+        return this.generateToken( user.id, user.username);
+
+    }
+
+    async changePassword(data: changePasswordDto, userId: number): Promise<any>{
```
Comment (DINESHKARPE) [inline src/auth/auth.service.ts]: why return any try Specific object

File: src/auth/auth.service.ts [hunk 3]
```diff
@@ -50,8 +56,64 @@ export class AuthService {
...
+
+        //hash new password
+        const newHashedPassword = await this.hashPassword(data.newPassword);
+        return this.userService.updatePassword(userId, newHashedPassword);
+    }
+
+    async forgotPassword(data: forgotPasswordDto): Promise<any>{
+        const user = await this.userService.findByEmail(data.email);
+
+        if(user){
```
Comment (DINESHKARPE) [inline src/auth/auth.service.ts]: if user not found then return user not found

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -50,8 +56,64 @@ export class AuthService {
...
+        //check valid user
+        const user = await this.userService.findById(validToken.user_id);
+
+        if(!user){
+            throw new InternalServerErrorException();
+        }
 
+        //change password
+        const hashedPassword = await this.hashPassword(data.newPassword);
+        return this.userService.updatePassword(validToken.user_id, hashedPassword);
```
Comment (DINESHKARPE) [inline src/auth/auth.service.ts]: require async because DB operation take time some time in service

File: src/auth/dto/change-password.dto.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { IsString, Matches, MinLength, matches } from "class-validator";
+
+export class  changePasswordDto{
```
Comment (DINESHKARPE) [inline src/auth/dto/change-password.dto.ts]: ChangePasswordDto - Use correct case

File: src/auth/dto/forgot-password.dto.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { IsEmail, IsNotEmpty, IsString } from "class-validator";
+
+export class forgotPasswordDto{
```
Comment (DINESHKARPE) [inline src/auth/dto/forgot-password.dto.ts]: Use correct case

File: src/auth/dto/reset-password.dto.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { IsString, Matches, MinLength, matches } from "class-validator";
+
+export class resetPasswordDto{
```
Comment (DINESHKARPE) [inline src/auth/dto/reset-password.dto.ts]: Use correct case

File: src/auth/dto/signIn-dto.ts [hunk 8]
```diff
@@ -1,9 +1,11 @@
-import { IsEmail, IsNotEmpty, IsString } from "class-validator";
+import { IsEmail, IsNotEmpty, IsString, Matches, MinLength } from "class-validator";
 
 export class signInDto{
```
Comment (DINESHKARPE) [inline src/auth/dto/signIn-dto.ts]: Use correct case

File: src/auth/refresh-token-service.ts [hunk 9]
```diff
@@ -9,12 +9,12 @@ export class RefreshTokenService{
     constructor(@InjectRepository(RefreshToken) private readonly refreshTokenRepository: Repository<RefreshToken>,
     private readonly configService: ConfigService, ){}
 
-    async create(token:  string, userId: number): Promise<void>{
+    async create(token:  string, user_id: number): Promise<void>{
         const expiredAfterDays = Number(this.configService.get('REFRESH_TOKEN_EXPIRY'));
```
Comment (DINESHKARPE) [inline src/auth/refresh-token-service.ts]: Create separate method

File: src/auth/reset-token.service.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+@Injectable()
+export class ResetTokenService{
+    constructor(
+        @InjectRepository(ResetToken) private readonly resetTokenRepository: Repository<ResetToken>,
+        private readonly configService: ConfigService
+    ){}
+
+    async create(user_id: number, token: string): Promise<void>{
+        const expiredAfterMinutes = Number(this.configService.get('RESET_TOKEN_EXPIRY'));
```
Comment (DINESHKARPE) [inline src/auth/reset-token.service.ts]: const expiredAfterMinutes = Number(this.configService.get('RESET_TOKEN_EXPIRY'));
        const expires_in = new Date();
        expires_in.setMinutes(expires_in.getMinutes() + expiredAfterMinutes); 

        Create Separate method

File: src/blogs/blogs.service.ts [hunk 11]
```diff
@@ -20,19 +20,46 @@ export class BlogsService {
...
+
     return await this.blogRepository.findAndCount({where : {user_id : user_id}});
   }
+
+  async searchBlogs(searchTerm: string){
+    if(!searchTerm || searchTerm.length < 3){
+      throw new BadRequestException("Search term must be at least 3 characters");
+    }
+  
+    return this.blogRepository.createQueryBuilder('blog')
```
Comment (DINESHKARPE) [inline src/blogs/blogs.service.ts]: In Return to many things

File: src/config/logger-config.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import * as winston from 'winston';
+import 'winston-daily-rotate-file';
+
+//ustom format for logs
+const customFormat = winston.format.printf(({ level, message, timestamp, context, ...meta }) => {
+    return `[${timestamp}] ${level.toUpperCase()}: ${message} ${meta.errorMessage ? `| Error: ${meta.errorMessage}` : ''} ${meta.statusCode ? `| Status: ${meta.statusCode}` : ''} ${meta.duration ? `| Duration: ${meta.duration}` : ''}`;
+});
+
+
+const logger = winston.createLogger({
```
Comment (DINESHKARPE) [inline src/config/logger-config.ts:11]: Load all value from property file and env file

File: src/services/mail.service.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+            auth : {
+                user: this.configService.get('EMAIL_USER'),
+                pass: this.configService.get('EMAIL_PASSWORD')
+            }
+
+        })
+    }
+
+    async sendPasswordResetEmail(to: string, token: string) {
```
Comment (DINESHKARPE) [inline src/services/mail.service.ts]: Don't use hadcode value , load html from template engine

**Review by DINESHKARPE — CHANGES_REQUESTED** [review]

File: src/auth/auth.service.ts (hunk 3 above)
Comment (imrahulmane) [reply src/auth/auth.service.ts]: This was done intentionally! While doing forgot password if any hacker is trying with different emails we don't want to tell user that "This user is not exists"

**Pull request merged**